http-client/Network/HTTP/Client/Request.hs

```diff
@@ -466,7 +475,8 @@ requestBuilder req Connection {..} = do
         | otherwise  = fromByteString "http://"
 
     requestHostname
-        | isJust (proxy req) && not (secure req)
+        | isJust (proxy req) && (
+            proxySecureMode req == ProxySecureSendRequest || not (secure req))
```

Comment:
The logic here is a little difficult to follow, and it's not resilient to possible increases in the number of `ProxySecureMode` data constructors in the future. Can we instead use more explicit pattern matching to hopefully clarify the meaning here, as well as cause a warning if an extra data constructor is added?